# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/gfxUserFontSet.cpp
# Commit: 97c97bbc0691
# Full Hash: 97c97bbc0691396e57374c7fd46214d379a02152
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2021-04-02 03:33:31
# Regressor Bug: 1694123
# File Overlap Count: 1
# Description:
#   Bug 1694123 - Ensure user fonts with src:local() are refreshed after a font-list update, even if the local font lookup previously failed. r=jwatt
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D106511
# ==============================================================================

diff -r 8badf00b02be -r 97c97bbc0691 gfx/thebes/gfxUserFontSet.cpp
--- a/gfx/thebes/gfxUserFontSet.cpp	Thu Apr 01 17:25:12 2021 +0000
+++ b/gfx/thebes/gfxUserFontSet.cpp	Thu Apr 01 17:25:19 2021 +0000
@@ -49,6 +49,7 @@
     : gfxFontEntry("userfont"_ns),
       mUserFontLoadState(STATUS_NOT_LOADED),
       mFontDataLoadingState(NOT_LOADING),
+      mSeenLocalSource(false),
       mUnsupportedFormat(false),
       mFontDisplay(aFontDisplay),
       mLoader(nullptr),
@@ -355,6 +356,13 @@
 }
 
 void gfxUserFontEntry::LoadNextSrc() {
+  // If LoadCanceled was called, we need to reset mSrcIndex so that all
+  // potential sources are re-considered.
+  if (mFontDataLoadingState == NOT_LOADING) {
+    mSrcIndex = 0;
+    mSeenLocalSource = false;
+  }
+
   NS_ASSERTION(mSrcIndex < mSrcList.Length(),
                "already at the end of the src list for user font");
   NS_ASSERTION((mUserFontLoadState == STATUS_NOT_LOADED ||
@@ -419,17 +427,20 @@
     if (currSrc.mSourceType == gfxFontFaceSrc::eSourceType_Local) {
       // Don't look up local fonts if the font whitelist is being used.
       gfxPlatformFontList* pfl = gfxPlatformFontList::PlatformFontList();
-      gfxFontEntry* fe =
-          pfl && pfl->IsFontFamilyWhitelistActive()
-              ? nullptr
-              : gfxPlatform::GetPlatform()->LookupLocalFont(
-                    currSrc.mLocalName, Weight(), Stretch(), SlantStyle());
-      nsTArray<gfxUserFontSet*> fontSets;
-      GetUserFontSets(fontSets);
-      for (gfxUserFontSet* fontSet : fontSets) {
-        // We need to note on each gfxUserFontSet that contains the user
-        // font entry that we used a local() rule.
-        fontSet->SetLocalRulesUsed();
+      gfxFontEntry* fe = nullptr;
+      if (!pfl->IsFontFamilyWhitelistActive()) {
+        fe = gfxPlatform::GetPlatform()->LookupLocalFont(
+            currSrc.mLocalName, Weight(), Stretch(), SlantStyle());
+        // Note that we've attempted a local lookup, even if it failed,
+        // as this means we are dependent on any updates to the font list.
+        mSeenLocalSource = true;
+        nsTArray<gfxUserFontSet*> fontSets;
+        GetUserFontSets(fontSets);
+        for (gfxUserFontSet* fontSet : fontSets) {
+          // We need to note on each gfxUserFontSet that contains the user
+          // font entry that we used a local() rule.
+          fontSet->SetLocalRulesUsed();
+        }
       }
       if (fe) {
         LOG(("userfonts (%p) [src %d] loaded local: (%s) for (%s) gen: %8.8x\n",
@@ -1059,11 +1070,15 @@
     for (const auto& f : fonts) {
       auto ufe = static_cast<gfxUserFontEntry*>(f.get());
       // If the user font entry has loaded an entry using src:local(),
-      // discard it as no longer valid, and reset the load state so that
-      // the load will be re-done based on the updated font list.
+      // discard it as no longer valid.
       if (ufe->GetPlatformFontEntry() &&
           ufe->GetPlatformFontEntry()->IsLocalUserFont()) {
         ufe->mPlatformFontEntry = nullptr;
+      }
+      // We need to re-evaluate the source list in the context of the new
+      // platform fontlist, whether or not the entry actually used a local()
+      // source last time, as one might be newly available.
+      if (ufe->mSeenLocalSource) {
         ufe->LoadCanceled();
       }
     }